### ISSUE TITLE
support for curried utility functions

### ISSUE THREAD

**Issue by sbenthall** [econ-ark/HARK#706]
The `utilities.CRRAutility` method takes a consumption parameter `c` and a risk aversion factor `gam`.

When used in practice, the `gam` function needs to be filled in... this is currently done with a lambda wrapper

https://github.com/econ-ark/HARK/blob/master/HARK/ConsumptionSaving/ConsIndShockModel.py#L410

It would be easy to implement a curried version of this, so that

```
self.u = makeCRRAutility(CRRA)
```
would work.

Could clean up this method renaming code in the process:
https://github.com/econ-ark/HARK/blob/master/HARK/ConsumptionSaving/ConsIndShockModel.py#L41-L47

**Comment by sbenthall** [comment]
Another area where these utility functions are recomposed is in the building of these value function objects:

https://github.com/econ-ark/HARK/blob/master/HARK/ConsumptionSaving/ConsIndShockModel.py#L148-L306

If I'm reading it right, these rather heavily comment HARKObjects are basically just functions compose of other functions and parameters.

Maybe I'm missing something about how these get used in a more object-oriented way?

**Comment by mnwhite** [comment]
How would you program this functionality?  I don't understand your
comment/criticism here.

I'm fine with curried functions, but don't understand what you're saying
about ValueFunction (et al).

**Comment by sbenthall** [comment]
I guess I'm just thrown off a bit by the coding style.

Looking at just this:
https://github.com/econ-ark/HARK/blob/master/HARK/ConsumptionSaving/ConsIndShockModel.py#L148-L190

This is ~40 lines of code. I see it's assiduously commented as is appropriate for the object-oriented style.

This is totally minor, but 

```
def makeValueFunc(vFuncNvrs, CRRA):
    '''
    Returns a value function.  The underlying interpolation is
    in the space of (m,u_inv(v)); this class "re-curves" to the value function.

     Parameters
     ----------
        vFuncNvrs : function
            A real function representing the value function composed with the
            inverse utility function, defined on market resources: u_inv(vFunc(m))
        CRRA : float
            Coefficient of relative risk aversion.
        Returns
        -------
       value_function: function
           Input: m : float or np.array
            Market resources (normalized by permanent income) whose value is to
            be found.
        Returns: v :  float or np.array
            Lifetime value of beginning this period with market resources m; has
            same size as input m.
    '''
    def value_func(m):   
        return utility(self.func(m),gam=self.CRRA)
    return value_func   
```

Is about half as long and is, for me at least, a bit easier to read because I don't have to think through both 'initialization' and 'calling'.

There's a lot of room for disagreement about this and maybe it's not worth mentioning. But I have a feeling that adopting a slightly more functional (as in, "functional programming", not "operational") style would make the code feel lighter weight. (This was used earlier in the makeOOSolver conversion)

**Comment by mnwhite** [comment]
I'm sorry, I still don't understand.  It doesn't look like that code does
anything?  It takes in arguments called vFuncNvrs and CRRA, and returns a
thing called value_func that refers to self, which has attributes func and
CRRA.  Were these supposed to be set as attributes of a new object
somewhere?

**Comment by sbenthall** [comment]
You are quite right. I've edited the code--i think i fixed the bug.

**Comment by mnwhite** [comment]
Ok, now I understand.  This won't work with HARK as is because it does not have a distance_criteria attribute, so it won't work with HARK's "recursive supnorm" distanceFunc, so it can't be used in an infinite horizon model (if vFunc is a convergence criterion).

distanceFunc knows how to handle floats, arrays, lists, and objects (that have a distance_criteria attribute) recursively.  It doesn't know how to handle lambdas or other functions.

**Comment by mnwhite** [comment]
Man that was a run-on sentence; really sorry.

**Comment by alanlujan91** [comment]
is this resolved by #1168 ?

**Comment by mnwhite** [comment]
18 months later: Yes, I think it was. We have a unified representation of utility functions.
